File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.cpp
class Solution {
public:
    int minSpeedOnTime(vector<int>& dist, double hour) {
        //Bin search
        const int N = dist.size();
        
        //search space
        long minSpeed = 1;
        long maxSpeed = INT_MAX;
        
        int minSpeedReq = INT_MAX;
        while(minSpeed <= maxSpeed) {
            long avgSpeed = minSpeed +(maxSpeed -minSpeed)/2;
            if(isPossibleToReach(dist, hour, avgSpeed)) {
                minSpeedReq = avgSpeed;
                maxSpeed = avgSpeed -1;
            } else {
                minSpeed = avgSpeed +1;
            }
        }
        
        return(minSpeedReq == INT_MAX)? -1 : minSpeedReq;
    }
    
private:
    bool isPossibleToReach(vector<int>& dist, double hour, long speed) {
        const int N = dist.size();
        double currTime = 0.0;
        for(int i = 0; i < N -1; i++) {
            int currDist = dist[i];
            currTime += ceil((double)currDist /speed);
            if(currTime > hour) return false;
        }
        currTime += ((double)dist[N -1] /speed);
        
        return (currTime <= hour);
    }
};

File: 1870-minimum-speed-to-arrive-on-time/1870-minimum-speed-to-arrive-on-time.java
class Solution {
    public int minSpeedOnTime(int[] dist, double hour) {
        
        //search space
        long low = 1, high = Integer.MAX_VALUE;
        int minSpeed = -1;
        
        while(low <= high) {
            long mid = low +(high -low) /2;
            if(isPossibleReach(dist, hour, mid)) {
                minSpeed = (int)mid;
                high = mid -1;
            } else {
                low = mid +1;
            }
        }
        
        return minSpeed;
    }
    
    private boolean isPossibleReach(int[] dist, double hour, long speed) {
        
        final int N = dist.length;
        double currTime = 0;
        
        for(int i = 0; i < N -1; i++) {
            currTime += Math.ceil((double)dist[i] /speed);
        }
        currTime += (double)dist[N -1] /speed;
        
        return (currTime <= hour);
    }
}